### PR TITLE
feat: Add Support for AzureDevops Webhooks

### LINKED ISSUE #3937: Add Support for AzureDevops Webhooks
# Summary

Add Support For Azure DevOps WebHooks. 

# Motivation

Github, Bitbucket, etc have support and Azure devops provides similar webhooks. 

https://docs.microsoft.com/en-us/azure/devops/service-hooks/events?view=azure-devops

# Proposal

Add Support for the GitPush event,https://docs.microsoft.com/en-us/azure/devops/service-hooks/events?view=azure-devops#git.push

as seen here, https://github.com/argoproj/argo-cd/blob/95820cf64fe695f0c30eedfc32b6de01a35b851c/util/webhook/webhook.go#L36 and sync associated applications.

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/3937

PR adds support for AzureDevops Webhooks

### CODE REVIEW

File: docs/operator-manual/webhook.md [hunk 1]
```diff
@@ -36,12 +36,14 @@ In the `argocd-secret` kubernetes secret, configure one of the following keys wi
...
 | Provider        | K8s Secret Key                   |
-|-----------------| ---------------------------------|
+|-----------------|----------------------------------|
 | GitHub          | `webhook.github.secret`          |
 | GitLab          | `webhook.gitlab.secret`          |
 | BitBucket       | `webhook.bitbucket.uuid`         |
 | BitBucketServer | `webhook.bitbucketserver.secret` |
 | Gogs            | `webhook.gogs.secret`            |
+| Azure DevOps    | `webhook.azuredevops.username`   |
+|                 | `webhook.azuredevops.password`   |
```
Comment (ocdi) [inline docs/operator-manual/webhook.md:46]: as a DevOps user I'd like a bit of clarity on what this password is - is it a PAT or something else?

Comment (ocdi) [reply docs/operator-manual/webhook.md:46]: Ah nevermind, I presume when adding the webhook you generate a password and this is to validate it on the argo end
image

File: util/settings/settings.go [hunk 2]
```diff
@@ -1576,6 +1590,12 @@ func (mgr *SettingsManager) SaveSettings(settings *ArgoCDSettings) error {
 		if settings.WebhookGogsSecret != "" {
 			argoCDSecret.Data[settingsWebhookGogsSecretKey] = []byte(settings.WebhookGogsSecret)
 		}
+		if settings.WebhookAzureDevOpsUsername != "" {
+			argoCDSecret.Data[settingsWebhookAzureDevOpsUsernameKey] = []byte(settings.WebhookAzureDevOpsUsername)
+		}
+		if settings.WebhookAzureDevOpsPassword != "" {
+			argoCDSecret.Data[settingsWebhookAzureDevOpsPasswordKey] = []byte(settings.WebhookAzureDevOpsPassword)
+		}
```
Comment (ishitasequeira) [inline util/settings/settings.go:1598]: Same comment here regarding a check on username and password

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
I don't have a way to test this, but the overall change looks good apart from couple of small comments.

File: docs/operator-manual/webhook.md (hunk 1 above)
Comment (crenshaw-dev) [reply docs/operator-manual/webhook.md:46]: This screenshot might be a good thing to add to docs. :-)

File: util/settings/settings.go [hunk 3]
```diff
@@ -1457,6 +1465,12 @@ func (mgr *SettingsManager) updateSettingsFromSecret(settings *ArgoCDSettings, a
 	if gogsWebhookSecret := argoCDSecret.Data[settingsWebhookGogsSecretKey]; len(gogsWebhookSecret) > 0 {
 		settings.WebhookGogsSecret = string(gogsWebhookSecret)
 	}
+	if azureDevOpsUsername := argoCDSecret.Data[settingsWebhookAzureDevOpsUsernameKey]; len(azureDevOpsUsername) > 0 {
+		settings.WebhookAzureDevOpsUsername = string(azureDevOpsUsername)
+	}
+	if azureDevOpsPassword := argoCDSecret.Data[settingsWebhookAzureDevOpsPasswordKey]; len(azureDevOpsPassword) > 0 {
+		settings.WebhookAzureDevOpsPassword = string(azureDevOpsPassword)
+	}
```
[the remark this answers is not in the document]
Comment (gdsoumya) [reply util/settings/settings.go:1473]: +1, I think both username and password are probably needed for basic auth, unless azure allows empty passwords for the webhook.

**Review by gdsoumya — APPROVED** [review]
LGTM, minor nit picks about env validation.

File: util/settings/settings.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexmt) [reply util/settings/settings.go:1473]: Both username and password are optional in azure so I think it has to be optional in Argo CD as well.

File: applicationset/webhook/webhook.go [hunk 4]
```diff
@@ -19,9 +19,9 @@ import (
 	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
 	argosettings "github.com/argoproj/argo-cd/v2/util/settings"
 
+	"github.com/go-playground/webhooks/v6/github"
```
[the remark this answers is not in the document]
Comment (alexmt) [reply applicationset/webhook/webhook.go:22]: I tried but looks like there was a change and new package name is `github.com/go-playground/webhooks/v6`. Here is `go get` output:

```
go get gopkg.in/go-playground/webhooks.v6
go: gopkg.in/go-playground/webhooks.v6@v6.2.0: parsing go.mod:
	module declares its path as: github.com/go-playground/webhooks/v6
	        but was required as: gopkg.in/go-playground/webhooks.v6
```

File: util/webhook/webhook.go [hunk 5]
```diff
@@ -41,21 +42,26 @@ type settingsSource interface {
 // https://github.com/shadow-maint/shadow/blob/master/libmisc/chkname.c#L36
 const usernameRegex = `[a-zA-Z0-9_\.][a-zA-Z0-9_\.-]{0,30}[a-zA-Z0-9_\.\$-]?`
 
-var _ settingsSource = &settings.SettingsManager{}
+var (
+	_                              settingsSource = &settings.SettingsManager{}
```
[the remark this answers is not in the document]
Comment (alexmt) [reply util/webhook/webhook.go:46]: correct, thanks for clarifying it @gdsoumya !

File: util/settings/settings.go (hunk 2 above)
Comment (alexmt) [reply util/settings/settings.go:1598]: Both username and password are optional in azure so I think it has to be optional in Argo CD as well.

File: docs/operator-manual/webhook.md (hunk 1 above)
Comment (alexmt) [reply docs/operator-manual/webhook.md:55]: agree :) created separate sections for Github and Azure DevOps and used this screenshot in Azure DevOps section.

File: docs/operator-manual/webhook.md [hunk 6]
```diff
@@ -79,6 +81,10 @@ stringData:
 
   # gogs server webhook secret
   webhook.gogs.secret: shhhh! it's a gogs server secret
+
+  # azuredevops username and password
```
[the remark this answers is not in the document]
Comment (alexmt) [reply docs/operator-manual/webhook.md:94]: Done. Added separate section for Azure DevOps with clarification about basic auth.

**Comment by alexmt** [comment]
Thanks for review @gdsoumya  and @ishitasequeira ! PTAL

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Comment by GuitaristForEver** [comment]
in which argocd version this would be? :) 
@alexmt @crenshaw-dev @ishitasequeira @gdsoumya 

Thanks

**Comment by ishitasequeira** [comment]
@GuitaristForEver This would be added to `v2.9`.

**Comment by GuitaristForEver** [comment]
Thanks! @ishitasequeira 

looking forward to it :)

**Comment by GuitaristForEver** [comment]
Hi just checking in on this , is this still planned for 2.9? Couldn't find it on [ArgoCD Roadmap](https://github.com/orgs/argoproj/projects/25/views/14) 

thanks

**Comment by ishitasequeira** [comment]
@GuitaristForEver the feature has been added to 2.9. It has been included as part of [v2.9.0-rc1](https://github.com/argoproj/argo-cd/releases/tag/v2.9.0-rc1) as well. The latest available release candidate is [v2.9.0-rc4](https://github.com/argoproj/argo-cd/releases/tag/v2.9.0-rc4) in case you want to test it out.

**Comment by GuitaristForEver** [comment]
Thanks @ishitasequeira , I've tested it and it works!

**Comment by AlexGS74** [comment]
I tested it in RC1 too and it doesn't work for me as the regex fix was done only for SSH part of the match. The Http repos are still not matching.

see https://github.com/argoproj/argo-cd/issues/15700

**Pull request merged**